### ISSUE TITLE
 `render()` and `render_site()` behaves differently in the presence of _site.yml.

### ISSUE THREAD

**Issue by royfrancis** [rstudio/rmarkdown#1757]
Say, I have files *index.Rmd*, *slide1.Rmd* and *_site.yml*

_index.Rmd_

```
---
title: Home
---

This is the home
```

_slide1.Rmd_

```
---
title: "Xaringan slide"
output:
  xaringan::moon_reader
---

Hello
```

__site.yml_

```
name: RMD test
navbar:
  title: RMD test
  right:
    - text: Home
      href: index.html
    - text: Slide
      href: slide1.html
output:
  rmarkdown::html_document
```

If I run `rmarkdown::render('slide1.Rmd')` (in the presence of *_site.yml*), the output is the document defined in *_site.yml*. The YAML definition in the Rmd file is ignored.

If I run `rmarkdown::render_site()`, the output of *slide1.Rmd* is a xaringan presentation (ie; document defined inside slide1.Rmd). As expected.

But in addition, I get this warning:

_Output created: _site/index.html
Warning message:
In file.rename(output, output_dest) :
  cannot rename file '/home/user/Desktop/test-rmd/slide1.html' to './_site/slide1.html', reason 'No such file or directory'_

Is there some way to make `render()` use only the YAML definition inside the file and not that from `_site.yml`? Is the warning something to worry about?

Issue on [StackOverflow](https://stackoverflow.com/questions/56064050/render-xaringan-rmd-when-using-site-yml-file).

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by royfrancis** [comment]
New argument needed in `render()` to override the default use of `_site,yml`.

**Comment by cderv** [comment]
Let me try to explain here the current behavior of **rmarkdown**. 

## TL;DR

`output_format` search order is 
1. Site file (e.g `_site.yml`)
2. Common output file (e.g `_output.yml`)
3. Rmd file YAML header. 

All are merged in a list in this order and when no `output_format` is provided, the first one defined in this order list will be used. In your example above, `html_document()`. 

This can be overridden by specifying the output format directly. e.g:
* render_site('slide1.Rmd', "xaringan::moon_reader") or 
* render('slide1.Rmd', "xaringan::moon_reader")

@yihui is this correct from your experience ? Is this the expected behavior ? 

More details below from my dive into this - as reference for later and learning note for myself at least.

## Current behavior 

What is described below is probably something that end up being the result of several years of updates regarding the behavior. I don't know what is the "normal" behavior. One must have in mind that R Markdown website were primarily designed, I believe, to work with only one format for all the documents (the one provided in `_site.yml`). Options for this formats for a specific doc could be modified in YAML header of specific file. I am not sure using different formats in one site is something well supported as the moment.

Anyway, let's dive into the current behavior. 

When `output_format` is not specified in render function, the first one will be used looking in order into `_site.yml` > `_output.yml` > file YAML header. 
* When `output_format = NULL`, look into all those in order
  https://github.com/rstudio/rmarkdown/blob/eaf6efc3fe6d39373ae0cee7ce3a034119bb47f5/R/output_format.R#L503
  https://github.com/rstudio/rmarkdown/blob/eaf6efc3fe6d39373ae0cee7ce3a034119bb47f5/R/output_format.R#L512
  First one is chosen
  https://github.com/rstudio/rmarkdown/blob/eaf6efc3fe6d39373ae0cee7ce3a034119bb47f5/R/output_format.R#L536
* When `output_format = "all"`, render all those format in order
  https://github.com/rstudio/rmarkdown/blob/eaf6efc3fe6d39373ae0cee7ce3a034119bb47f5/R/output_format.R#L661
  https://github.com/rstudio/rmarkdown/blob/eaf6efc3fe6d39373ae0cee7ce3a034119bb47f5/R/output_format.R#L672

I think this order for searching format when none is provided is here since a long time. (but this does not mean we can't change I guess)

This also explains the behavior you see between `render()` and `render_site()` because: 
* `render(<input>)` defaults to `output_format = NULL` and take the first found
* `render_site()` defaults to `output_format = "all"` meaning it will render all formats found
* `render_site(<input>)` defaults to `output_format = "all"`, but changed to NULL because we are in incremental mode (render only one format) and as with `render()` the first one found will be used.

Any of this behavior can be overridden by specifying the `output_format`. A few example with your site: 

* `render_site()` will render to all formats and move to `_site` folder
  * `index.Rmd` to `html_document()`
  * `slide1.Rmd`to `html_document()` then to `xaringan::moon_reader()` - the later will overwrite the former because there are both HTML doc
* `rmarkdown::render('slide1.Rmd')` does not specify format so first one will be used, and `_site.yml` will be the one so `html_document()`. Moreover the file will be written the same folder as `Slide1.Rmd` and not `_site.Rmd`. 
*  `rmarkdown::render('slide1.Rmd', "xaringan::moon_reader")` will render to the format with options from the YAML header of the file. 
* In a website `render_site('slide1.Rmd')` is the function to use to render on file. By default, it will be the same as `rmarkdown::render('slide1.Rmd')` using `html_document()`
* Using `render_site('slide1.Rmd', "xaringan::moon_reader")` would render to the correct format.

## About overriding default

Currently, the way to override default behavior regarding output format is to use the `output_format` argument using a string corresponding to the YAML key of one of output filed in one of the files `_site.yml`, `_output.yml` or Rmd file YAML header.

## Thoughts ?

* I am note sure to see why we need to use `output_format: "all"` in `render_site()` 🤔 
* I think the search order for default format is here since a long time. I believe this is mainly in order to be able to modify specific options from a common output format from within a single Rmd file. (merging behavior). Not sure reversing it would be ok.
* Would something like this be a good new feature ? 
  Specify which output format to use as default by specifying where: 
    * `rmarkdown::render(output_format = "file_header")` -> use YAML header only
    * `rmarkdown::render(output_format = "file_output")` -> use `_output.yml` only
    * `rmarkdown::render(output_format = "file_site")` -> use `_site.yml` only
  I am really not sure this is better than specifying the desired output format explicitly by its name.


## Reproducing the behavior with R code

<details>
<summary>Here is a reprex to illustrate </summary>

````r
# setup project
dir.create(tmp_dir <- tempfile())
owd <- setwd(tmp_dir)

# create index
xfun::write_utf8(c(
  "---",
  "title: Home",
  "---",
  "",
  "This is the home",
  ""
), "index.Rmd")

# create slide
xfun::write_utf8(c(
  "---",
  "title: 'Xaringan slide'",
  "output: xaringan::moon_reader",
  "---",
  "",
  "Hello",
  ""
), "slide1.Rmd")

# create _site.yml
yaml::write_yaml(list(
  name = "RMD test",
  navbar = list(title = "RMD test", right = list(
    list(text = "Home", href = "index.html"),
    list(text = "Slide",
         href = "slide1.html")
  )),
  output = "rmarkdown::html_document"
), "_site.yml")

# check files 
ori_files <- list.files()
ori_files

# Render website - We'll have a xaringan presentation
# because all output format will be rendered 
# and xaringan::moon_reader() is the last
rmarkdown::render_site()

# serve website
servr::httd("_site")

 
# re render 1 file - We are in incremental mode for rmarkdown website
# Only one format will be render and the one in _site.yml takes precedence. have only
rmarkdown::render_site("slide1.Rmd")

# We need to specify the format for this one
rmarkdown::render_site("slide1.Rmd", "xaringan::moon_reader")

# We can also use render but we'll use the benefit of website rendering 
# (moving file to _site folder for example)
# It works as render_site where precedence is _site.yml > _output.yml > yaml header

# we'll have a `html_document()` output in ./slide1.html
rmarkdown::render("slide1.Rmd") 
browseURL("./slide1.html")

# we can overide precedence using format name
rmarkdown::render("slide1.Rmd", "xaringan::moon_reader") 
browseURL("./slide1.html")

# cleaning 
setwd(owd)
unlink(tmp_dir, recursive = TRUE)
````

</details>

**Comment by royfrancis** [comment]
@cderv Thank you for the detailed analyses :) This is a clear explanation!

I never paid much attention to `output_format` argument in the function as I always specified it in the Rmd yaml. In a website context, I just assumed that the output format was whatever indicated in `_site.yml`. This makes sense because most Rmd/html files in a website would share the same output/styles etc and there is no need to specify output in every Rmd.

Now, if I want the style/format of one of the pages to be different, I would manually specify an `output_format` in that Rmd file alone. And that is now prioritized over `_site.yml`. This all works as expected intuitively.

The issue was that `rmarkdown::render()` wouldn't use the output specified in the Rmd and instead take from `_site.yml`. Renaming `_site.yml` temporarily was a bit annoying.

As you mentioned, `rmarkdown::render("slide1.Rmd", "xaringan::moon_reader")` solves this issue for now. But, it would be nice to be able to say, use whatever is in the yaml header.

Maybe the issue is that search order is implicit. In a scenario where file_header, file_output and file_site are all available, it would be nice to be able to specify/change/set the order or source of output definitions explicitly.

**Comment by yihui** [comment]
I feel it makes more sense for the `output` format specified in the YAML header to have higher precedence than the format in `_site.yml`. The latter is global, and the former is local. In general, local options should override global options, but not the other way around.

**Comment by cderv** [comment]
Yes that suprised me too. 

Currently, yaml header has the higher precedence regarding option sets in it. It will override the option defined in site.yaml

This behavior above should stay. But I think, if it is not the same format in yaml header than in site yaml, and no format is specified in `output format` then it should be chosen as first format. 

I think it should have precedence this way. it feels like a pretty big change though but maybe it is fine. 

The `output format = "all"` in `render_site()` is puzzling to me also. But that is another topic.
